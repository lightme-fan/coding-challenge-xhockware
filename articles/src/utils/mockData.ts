export const article = {
    id: "123",
    source: {
    id: "cnbc",
    name: "CNBC"
  },
    author: "Amelia Lucas",
    title: "Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC",
    description: "Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.",
    url: "https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html",
    urlToImage: "https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846",
    publishedAt: "2019-10-18T10:42:46Z",
    content: "Coca-Cola on Friday reported quarterly revenue.",
    viewText: "Read"
}