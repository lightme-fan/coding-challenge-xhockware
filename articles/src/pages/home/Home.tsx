import React, { useEffect, useState } from 'react';
import { ArticleDescription, ArticleListItem } from '../../components';
import styles from "./styles.module.css";
import { v4 as uuid } from 'uuid';
import { Article } from '../../utils/types';

const { articles_container, bar, left_list, parent_list, article_list, article_details } = styles;

const Home = () => {
  const uniqueId: string = uuid();
  const [articles, setArticles] = useState<Article[] | null | any>(null);
  const [articleDetails, setArticleDetails] = useState<Article | null | any>(null);

  const fetchArticles = async () => {
    try {
      const response = await fetch('http://localhost:8000/v1/news?q=');
      const jsonData = await response.json();
      const modifiedArticles = jsonData.articles.map((article: Article, index: number) => {
        return {
          ...article,
          id: `${index}${uniqueId}`,
          viewText: "Unread"
        };
      })
      setArticles(modifiedArticles);
    } catch (error) {
      return error
    }
  };

  useEffect(() => {
    fetchArticles();
    // eslint-disable-next-line
  }, []);

  const handleClickOnArticle = (id: string) => {
    const foundArticle = articles && articles.find((article: any) => article.id === id);
    const modifiedArticles = articles && articles.map((article: Article, index: number) => {
      if (article === foundArticle) {
        return {...article, viewText: "Read"};
      }
      return article;
    })

    setArticleDetails(foundArticle);
    setArticles(modifiedArticles)
  }

  return (
    <div className={articles_container}>
      <div className={parent_list} data-testid="left_side">
        <ul className={article_list}>
          {articles && articles?.map((article: Article) => {
            return <ArticleListItem
              key={article.id}
              article={article}
              handleClickOnArticle={() => handleClickOnArticle(article.id)}
            />
          })}
        </ul>
      </div>
      <div className={bar} data-testid="bar"></div>
      <div className={article_details} data-testid="right_side">
        <ArticleDescription article={articleDetails} />
      </div>
    </div>
  );
}

export default Home;
