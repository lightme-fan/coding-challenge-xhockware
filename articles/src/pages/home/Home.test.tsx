import '@testing-library/jest-dom'
import { render, waitFor } from '@testing-library/react'
import { article } from '../../utils/mockData'
import Home from './Home'

describe("Tests Home page ", () => {
  const mockArticle = [article];
  
  test("tests fetch articles function", async () => {
    global.fetch = jest.fn().mockImplementation(() =>
      Promise.resolve({
        json: jest.fn().mockResolvedValue(mockArticle),
      })
    );
    await waitFor(() => {
      expect(mockArticle).toHaveLength(mockArticle.length);
    });
  })

  test('tests if a list of article is display on the letf side, bar and right side', async () => {
    const component = render(<Home />);
    const { getByTestId } = component;
    expect(getByTestId("left_side")).toHaveClass("parent_list");
    expect(getByTestId("bar")).toHaveClass("bar");
    expect(getByTestId("right_side")).toHaveClass("article_details");
  })
})