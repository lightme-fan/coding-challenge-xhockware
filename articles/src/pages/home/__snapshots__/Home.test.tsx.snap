// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests Home page  tests if a list of article is display on the letf side 1`] = `
<ArticleListItem
  article={
    Object {
      "author": "Amelia Lucas",
      "content": "Coca-Cola on Friday reported quarterly revenue.",
      "description": "Coca-Cola on Friday reported quarterly earnings that met analysts' expectations.",
      "id": "123",
      "publishedAt": "2019-10-18T10:42:46Z",
      "source": Object {
        "id": "cnbc",
        "name": "CNBC",
      },
      "title": "Coca-Cola stock jumps as strong sales of Coke Zero Sugar continue to drive revenue growth - CNBC",
      "url": "https://www.cnbc.com/2019/10/18/coca-cola-ko-earnings-q3-2019.html",
      "urlToImage": "https://image.cnbcfm.com/api/v1/image/105375236-1533317591071rtx6drjw.jpg?v=1571332846",
      "viewText": "Read",
    }
  }
  handleClickOnArticle={[Function]}
/>
`;

exports[`Tests Home page  tests if a list of article is display on the letf side, bar and right side 1`] = `
<div
  class="parent_list"
  data-testid="left_side"
>
  <ul
    class="article_list"
  />
</div>
`;
